wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt

```diff
@@ -40,4 +52,221 @@ class Proto3WireProtocCompatibilityTests {
+    assertThat(jsonAdapter.fromJson(json)).isEqualTo(pizzaDelivery)
+  }
+
+  @Test fun wireProtocJsonRoundTrip() {
+    val protocMessage = PizzaOuterClass.PizzaDelivery.newBuilder()
+        .setAddress("507 Cross Street")
+        .addPizzas(PizzaOuterClass.Pizza.newBuilder()
+            .addToppings("pineapple")
+            .addToppings("onion")
+            .build())
+        .setPromotion(Any.pack(PizzaOuterClass.BuyOneGetOnePromotion.newBuilder()
+            .setCoupon("MAUI")
+            .build()))
+        .build()
+
+    val typeRegistry = JsonFormat.TypeRegistry.newBuilder()
+        .add(PizzaOuterClass.BuyOneGetOnePromotion.getDescriptor())
+        .add(PizzaOuterClass.FreeGarlicBreadPromotion.getDescriptor())
+        .build()
+
+    val jsonPrinter = JsonFormat.printer()
+        .usingTypeRegistry(typeRegistry)
+    val protocMessageJson = jsonPrinter.print(protocMessage)
+
+    val wireMessage = PizzaDelivery(
+        address = "507 Cross Street",
+        pizzas = listOf(Pizza(toppings = listOf("pineapple", "onion"))),
+        promotion = AnyMessage.pack(BuyOneGetOnePromotion(coupon = "MAUI"))
+    )
+
+    val moshi = Moshi.Builder()
+        .add(WireJsonAdapterFactory()
+            .plus(BuyOneGetOnePromotion.ADAPTER, FreeGarlicBreadPromotion.ADAPTER))
+        .build()
+    val jsonAdapter = moshi.adapter(PizzaDelivery::class.java)
+    val moshiMessageJson = jsonAdapter.toJson(wireMessage)
+
+    // Parsing the two json because this should be order insensitive.
+    assertJsonEquals(moshiMessageJson, protocMessageJson)
+
+    // Now each parses the other's json.
+    val jsonParser = JsonFormat.parser().usingTypeRegistry(typeRegistry)
+    val protocMessageDecodedFromWireJson = PizzaOuterClass.PizzaDelivery.newBuilder()
+        .apply { jsonParser.merge(moshiMessageJson, this) }
+        .build()
+
+    val wireMessageDecodedFromProtocJson = jsonAdapter.fromJson(protocMessageJson)
+
+    assertThat(protocMessageDecodedFromWireJson).isEqualTo(protocMessage)
+    assertThat(wireMessageDecodedFromProtocJson).isEqualTo(wireMessage)
+  }
+
+  @Test fun unregisteredTypeOnReading() {
+    val moshi = Moshi.Builder()
+        .add(WireJsonAdapterFactory())
+        .build()
+
+    val jsonAdapter = moshi.adapter(PizzaDelivery::class.java)
+
+    val json = """
+        |{
+        |  "address": "507 Cross Street",
+        |  "pizzas": [
+        |    {
+        |      "toppings": [
+        |        "pineapple",
+        |        "onion"
+        |      ]
+        |    }
+        |  ],
+        |  "promotion": {
+        |    "@type": "type.googleapis.com/squareup.proto3.pizza.BuyOneGetOnePromotion",
+        |    "coupon": "MAUI"
+        |  }
+        |}
+        """.trimMargin()
+
+    try {
+      jsonAdapter.fromJson(json)
+      fail()
+    } catch (expected: IOException) {
+      assertThat(expected).hasMessage("Cannot resolve type: " +
+          "type.googleapis.com/squareup.proto3.pizza.BuyOneGetOnePromotion")
```

Comment:
include the path in the expected message

Reply:
fixed

Reply:
fixed?